File: ios/mapbox_maps_flutter/Sources/mapbox_maps_flutter/Classes/Snapshotter/SnapshotterController.swift
import Foundation
@_spi(Experimental) import MapboxMaps
import Flutter

final class SnapshotterController: _SnapshotterMessenger {
    enum Error: Swift.Error {
        case imageConversionFailed
    }

    private var snapshotter: Snapshotter!
    private let eventHandler: MapboxEventHandler

    init(snapshotter: Snapshotter, eventTypes: [Int], binaryMessenger: FlutterBinaryMessenger, channelSuffix: String) {
        self.snapshotter = snapshotter

        eventHandler = MapboxEventHandler(
            eventProvider: snapshotter,
            binaryMessenger: binaryMessenger,
            eventTypes: eventTypes,
            channelSuffix: channelSuffix
        )
    }

    func getCameraState() throws -> CameraState {
        return snapshotter.cameraState.toFLTCameraState()
    }

    func setCamera(cameraOptions: CameraOptions) throws {
        snapshotter.setCamera(to: cameraOptions.toCameraOptions())
    }

    func setSize(size: Size) throws {
        snapshotter.snapshotSize = size.toCGSize()
    }

    func getSize() throws -> Size {
        return snapshotter.snapshotSize.toFLTSize()
    }

    func start(completion: @escaping (Result<FlutterStandardTypedData?, any Swift.Error>) -> Void) {
        snapshotter.start(overlayHandler: nil) { result in
            switch result {
            case .success(let image):
                if let imageData = image.pngData() {
                    completion(.success(.init(bytes: imageData)))
                } else {
                    completion(.failure(Error.imageConversionFailed))
                }
            case .failure(let failure):
                completion(.failure(failure))
            }
        }
    }

    func cancel() throws {
        snapshotter.cancel()
    }

    func coordinateBounds(camera: CameraOptions) throws -> CoordinateBounds {
        return snapshotter.coordinateBounds(for: camera.toCameraOptions()).toFLTCoordinateBounds()
    }

    func camera(coordinates: [Turf.Point], padding: MbxEdgeInsets?, bearing: Double?, pitch: Double?) throws -> CameraOptions {
        return snapshotter.camera(
            for: coordinates.map(\.coordinates),
            padding: padding?.toUIEdgeInsets(),
            bearing: bearing,
            pitch: pitch
        ).toFLTCameraOptions()
    }

    func tileCover(options: TileCoverOptions) throws -> [CanonicalTileID] {
        return snapshotter.tileCover(for: options.toTileCoverOptions())
            .map { $0.toFLTCanonicalTileID() }
    }

    func clearData(completion: @escaping (Result<Void, any Swift.Error>) -> Void) {
        do {
            snapshotter.clearData { error in
                if let error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
        }
    }
}

File: mapbox_maps_flutter_mobile/android/src/main/kotlin/com/mapbox/maps/mapbox_maps/snapshotter/SnapshotterController.kt
package com.mapbox.maps.mapbox_maps.snapshot

import android.content.Context
import android.graphics.Bitmap
import com.mapbox.geojson.Point
import com.mapbox.maps.Snapshotter
import com.mapbox.maps.mapbox_maps.MapboxEventHandler
import com.mapbox.maps.mapbox_maps.pigeons.CameraOptions
import com.mapbox.maps.mapbox_maps.pigeons.CameraState
import com.mapbox.maps.mapbox_maps.pigeons.CanonicalTileID
import com.mapbox.maps.mapbox_maps.pigeons.CoordinateBounds
import com.mapbox.maps.mapbox_maps.pigeons.MbxEdgeInsets
import com.mapbox.maps.mapbox_maps.pigeons.TileCoverOptions
import com.mapbox.maps.mapbox_maps.pigeons._SnapshotterMessenger
import com.mapbox.maps.mapbox_maps.toCameraOptions
import com.mapbox.maps.mapbox_maps.toCameraState
import com.mapbox.maps.mapbox_maps.toEdgeInsets
import com.mapbox.maps.mapbox_maps.toFLTCameraOptions
import com.mapbox.maps.mapbox_maps.toFLTCanonicalTileID
import com.mapbox.maps.mapbox_maps.toFLTCoordinateBounds
import com.mapbox.maps.mapbox_maps.toFLTSize
import com.mapbox.maps.mapbox_maps.toSize
import com.mapbox.maps.mapbox_maps.toTileCoverOptions
import java.io.ByteArrayOutputStream

class SnapshotterController(
  private val context: Context,
  private val snapshotter: Snapshotter,
  private val eventHandler: MapboxEventHandler
) : _SnapshotterMessenger {
  override fun getSize(): com.mapbox.maps.mapbox_maps.pigeons.Size {
    return snapshotter.getSize().toFLTSize(context)
  }

  override fun setSize(size: com.mapbox.maps.mapbox_maps.pigeons.Size) {
    snapshotter.setSize(size.toSize(context))
  }

  override fun getCameraState(): CameraState {
    return snapshotter.getCameraState().toCameraState(context)
  }

  override fun setCamera(cameraOptions: CameraOptions) {
    snapshotter.setCamera(cameraOptions.toCameraOptions(context))
  }

  override fun start(callback: (Result<ByteArray?>) -> Unit) {
    snapshotter.start(null) { snapshot, error ->
      if (snapshot != null) {
        val byteArray = ByteArrayOutputStream().also { stream ->
          snapshot.compress(Bitmap.CompressFormat.PNG, 100, stream)
        }.toByteArray()

        callback(Result.success(byteArray))
      } else {
        callback(Result.failure(Throwable(error ?: "Unknown error")))
      }
    }
  }

  override fun cancel() {
    snapshotter.cancel()
  }

  override fun coordinateBounds(camera: CameraOptions): CoordinateBounds {
    return snapshotter.coordinateBoundsForCamera(camera.toCameraOptions(context)).toFLTCoordinateBounds()
  }

  override fun camera(
    coordinates: List<Point>,
    padding: MbxEdgeInsets?,
    bearing: Double?,
    pitch: Double?
  ): CameraOptions {
    // TODO: remove default values from padding, bearing and pitch after they are optional on Android
    return snapshotter.cameraForCoordinates(
      coordinates,
      (padding ?: MbxEdgeInsets(0.0, 0.0, 0.0, 0.0)).toEdgeInsets(context),
      bearing ?: 0.0,
      pitch ?: 0.0
    )
      .toFLTCameraOptions(context)
  }

  override fun tileCover(options: TileCoverOptions): List<CanonicalTileID> {
    return snapshotter.tileCover(options.toTileCoverOptions(), null).map { it.toFLTCanonicalTileID() }
  }

  override fun clearData(callback: (Result<Unit>) -> Unit) {
    Snapshotter.clearData() {
      if (it.isError) {
        callback(Result.failure(Throwable(it.error)))
      } else {
        callback(Result.success(Unit))
      }
    }
  }
}